sqrl-openai/src/main/java/com/datasqrl/openai/extract_json.java

```diff
@@ -29,8 +38,21 @@ public String eval(String prompt, String modelName, Double temperature) {
     }
 
     public String eval(String prompt, String modelName, Double temperature, Double topP) {
-        return executeWithRetry(
-                () -> openAICompletions.callCompletions(prompt, modelName, true, null, temperature, topP)
+        if (prompt == null || modelName == null) return null;
```

Comment:
My suggestion is to move the executeWithRetry method into the FunctionMetricTracker so that you can make the code DRYer. In fact, I think we can combine those two helper classes. Whenever you execute some function with retry, you also want to count how many times you executed it and how many failures you have. Keeping these two helper clases orthogonal, makes it awkward to use and leads to more code.

Reply:
Thanks @mbroecheler! I was thinking about improving it further, but my understanding was that the retry utility is a temporary thing until we start using asynchronous functions. Does it make sense to improve it further if we are going to replace it anyway?

Reply:
@nandorsoma, yes, we will move to async eventually. That's why it would be nice to have all code related to retries and metrics on function invocations encapsulated in one class. Then we would only have to update the helper class and not the body of every function implementation.
For now, it mostly gives us better architectural encapsulation as we are learning how to implement functions.